mlx/fast.cpp

```diff
@@ -859,4 +864,229 @@ array affine_dequantize(
+  };
+  std::vector<std::pair<std::string, std::string>> attrs;
+  for (auto [attr, dtype] : metal_attributes) {
+    if (source_.find(attr) != std::string::npos) {
+      attrs.push_back({attr, dtype});
+    }
+  }
+
+  int index = 0;
+  constexpr int max_constant_array_size = 8;
+  // Add inputs
+  for (auto& [name, arr] : inputs) {
+    auto dtype = get_type_string(arr.dtype());
+    bool is_constant =
+        arr.is_available() && arr.size() < max_constant_array_size;
+    std::string location = is_constant ? "constant" : "device";
+    std::string ref = arr.ndim() == 0 ? "&" : "*";
+    kernel_source << "  const " << location << " " << dtype << ref << " "
+                  << name << " [[buffer(" << index << ")]]," << std::endl;
+    index++;
+    // Add input shape, strides and ndim
+    if (arr.ndim() > 0) {
+      kernel_source << "  const constant int* " << name << "_shape [[buffer("
+                    << index << ")]]," << std::endl;
+      index++;
+      kernel_source << "  const constant size_t* " << name
+                    << "_strides [[buffer(" << index << ")]]," << std::endl;
+      index++;
+      kernel_source << "  const constant int& " << name << "_ndim [[buffer("
+                    << index << ")]]," << std::endl;
+      index++;
+    }
+  }
+  // Add outputs
+  for (auto [name, dtype] : output_dtypes_) {
+    kernel_source << "  device " << get_type_string(dtype) << "* " << name
+                  << " [[buffer(" << index << ")]]";
+    if (index < inputs.size() + output_shapes_.size() - 1 || attrs.size() > 0) {
+      kernel_source << "," << std::endl;
+    } else {
+      kernel_source << ") {" << std::endl;
+    }
+    index++;
+  }
+  // Add metal attributes e.g. `threadgroup_index_in_grid`
+  for (auto [attr, dtype] : attrs) {
+    kernel_source << "  " << dtype << " " << attr << " [[" << attr << "]]";
+    if (index < attrs.size() - 1) {
+      kernel_source << "," << std::endl;
+    } else {
+      kernel_source << ") {" << std::endl;
+    }
+  }
+  kernel_source << source_ << std::endl;
+  kernel_source << "}" << std::endl;
+}
+
+std::string MetalKernel::write_template() {
+  std::ostringstream template_def;
+  template_def << "<";
+  int i = 0;
+  for (auto [name, arg] : template_args) {
+    if (i > 0) {
+      template_def << ", ";
+    }
+    if (std::holds_alternative<int>(arg)) {
+      template_def << std::get<int>(arg);
+    } else if (std::holds_alternative<bool>(arg)) {
+      template_def << std::get<bool>(arg);
+    } else if (std::holds_alternative<Dtype>(arg)) {
+      template_def << get_type_string(std::get<Dtype>(arg));
+    }
+    i++;
+  }
+  template_def << ">";
+  return template_def.str();
+}
+
+std::unordered_map<std::string, array> MetalKernel::run(
+    std::unordered_map<std::string, array>& inputs,
+    StreamOrDevice s_) {
+  auto s = to_stream(s_);
+  if (s.device != Device::gpu) {
+    throw std::invalid_argument("MetalKernel only works on GPU.");
```

Comment:
Nit: same comment re including the function / Class::method name in `[...]` as a prefix.